peps/pep-0776.rst

```diff
@@ -0,0 +1,794 @@
+
+Python uses the NODEFS approach when run with node and the ``ZipImporter``
+approach for the web example. We will continue with this approach.
+
+
+-----------------------------
+Console and Interactive Usage
+-----------------------------
+
+``stdin`` defaults to always returning ``EOF``, while ``stdout`` and ``stderr``
+default to calling ``console.log`` and ``console.error`` respectively. It is
+possible to pass handlers to ``bootstrapEmscriptenExecutable()`` to configure
+the standard streams, but no matter what the I/O devices have undesirable line
+buffering behavior that forces a new line when flushed. To implement a well
+behaved tty in browser, it is necessary to remove the default I/O devices and
+replace them in a ``preRun`` hook.
+
+Making ``stdin`` work correctly in the browser poses an additional challenge
+because it is not allowed to block for user input in the main thread of the
+browser. If Emscripten is run in a webworker and served with the shared memory
+headers, it is possible to receive input using shared memory and atomics. It is
+also possible for a ``stdin`` device to block in a simpler and more efficient
+manner using stack switching using the experimental JavaScript Promise
+Integration API.
+
+Pyodide replaces the standard I/O devices in order to fix the line buffering
+behavior. When Pyodide is run in node, ``stdin``, ``stdout``, and ``stderr`` are
+by default connected to ``process.stdin``, ``process.stdout``, and
+``process.stderr`` and so the standard streams work as a tty out of the box.
+Pyodide also ensures that ``shutil.get_terminal_size`` returns results
+consistent with ``process.stdout.rows`` and ``process.stdout.columns``. Pyodide
+currently has no support for stack switching ``stdin``.
+
+Currently, the Emscripten Python node runner uses the default I/O that
+Emscripten provides. The web example uses ``Atomics`` for ``stdin`` and has
+custom ``stdout`` and ``stderr`` handlers, but they exhibit the undesirable line
+buffering behavior. We will upstream the standard streams behaviors from
+Pyodide.
+
+In the long term, we hope to implement stack switching ``stdin`` devices, but
+that is out of scope for this PEP.
+
+
+-----------------
+Dynamic Libraries
+-----------------
+
+Main Thread Synchronous Loading Limit
+=====================================
+
+In the main browser thread, a dynamic library can only be loaded synchronously
+if it is at most 4 kilobytes. This excludes most nontrivial dynamic libraries.
+This limit is not present in Node and can be avoided by using a web worker. If
+stack switching is available, then it is possible to make ``dlopen()`` stack
+switch in order to instantiate a dynamic library synchronously.
+
+To avoid the synchronous loading limit, Pyodide currently preloads all dynamic
+libraries present in a wheel when installing the wheel (or on startup). This is
+a significant disadvantage with pages like scipy that include a very large
```

Comment:
"libraries" like scipy, not pages?

Reply:
maybe `packages` is the most appropriate word here.